usr/src/utile-1.1/setserial.8
.\" Copyright 1992, 1993 Rickard E. Faith (faith@cs.unc.edu)
.\" May be distributed under the GNU General Public License
.TH SETSERIAL 8 "16 January 1993" "Linux 0.99" "Linux Programmer's Manual"
.SH NAME
setserial \- get/set Linux serial port information
.SH SYNOPSIS
.B /etc/setserial
.B "[ \-g ]"
serial-device
.BR "[ " opt1 " [ " arg " ] ] [ " opt2 " [ " arg " ] ] ..."
.SH DESCRIPTION
Without any options,
.B setserial
will print out the port type (i.e., 8250, 16450, 16550, 16550A), the
hardware I/O port, the hardware IRQ line, the "baud base," and the "flags."
The
.I device
parameter will usually have the following form:
.BR /dev/ttys[0-3] .

The "standard MS-DOS" port associations are given below:

.nf
.RS
/dev/ttys0 (COM1), port 0x3f8, irq 4
/dev/ttys1 (COM2), port 0x2f8, irq 3
/dev/ttys2 (COM3), port 0x3e8, irq 4
/dev/ttys3 (COM4), port 0x2e8, irq 3
.RE
.fi

Note, however, that Linux is not able to have a single IRQ line shared
between multiple devices (special serial port hardware, such as the AST
FourPort, can be used to overcome this limitation).  Alternatively, some
serial port cards (such as the Accent Async and the Twincom ACI/550) allow
the selection of alternative I/O port addresses and IRQ lines.  For
example, the Linux 0.99.1 kernel is set up to check the following
configurations if the
.I CONFIG_ACCENT_ASYNC
option is selected when building the kernel
.RI ( "/dev/ttys4" " through " "/dev/ttys11"
are reserved for the AST FourPort):

.nf
.RS
/dev/ttys12, port 0x330, irq 4
/dev/ttys13, port 0x338, irq 4
.RE
.fi

If the kernel's auto-detection of I/O port addresses and/or IRQ lines does
not work reliably on your system, the
.B setserial
program can be called from your
.I /etc/rc.local
to adjust the configuration at boot time.  The selection of an alternative
IRQ line is difficult, since most of them are already used.  The following
table lists the "standard MS-DOS" assignments of available IRQ lines:

.nf
.RS
IRQ 3: COM2
IRQ 4: COM1
IRQ 5: LPT2
IRQ 7: LPT1
.RE
.fi

Most people find that IRQ 5 is a good choice, assuming that there is only
one parallel port active in the computer.  The only other available IRQ
lines are 3, 4, and 7, and these are probably used by the other serial and
parallel ports.  (If your serial card has a 16bit card edge connector, and
supports higher interrupt numbers, then IRQ 10, 11, 12, and 15 are also
available.)

On AT class machines, IRQ 2 is seen as IRQ 9, and Linux will interpret it
in this manner.

IRQ's other than 2 (9), 3, 4, 5, 7, 10, 11, 12, and 15, should
.I not
be used, since they are assigned to other hardware and cannot, in general,
be changed.  Here are the "standard" assignments:

.nf
.RS
IRQ  0      Timer channel 0
IRQ  1      Keyboard
IRQ  2      Cascade for controller 2
IRQ  3      Serial port 2
IRQ  4      Serial port 1
IRQ  5      Parallel port 2 (Reserved in PS/2)
IRQ  6      Floppy diskette
IRQ  7      Parallel port 1
IRQ  8      Real-time clock
IRQ  9      Redirected to IRQ2
IRQ 10      Reserved
IRQ 11      Reserved
IRQ 12      Reserved (Auxillary device in PS/2)
IRQ 13      Math coprocessor
IRQ 14      Hard disk controller
IRQ 15      Reserved
.RE
.fi

.SH OPTIONS
All argument values are assumed to be in decimal unless preceeded by "0x".

If the
.B \-g
option is given, a list of devices can be specified, and the
characteristics of those devices will be printed.  Without the
.B \-g
option, the characteristics of single device can be printed, a list is not
allowed.

.TP
.BR port " port_number"
The
.B port
option sets the I/O port, as described above.
.TP
.BR irq " irq_number"
The
.B irq
option sets the hardware IRQ, as described above.
.TP
.BR uart " uart_type"
This option is used to set the UART type.  The permitted types are none,
8250, 16450, 16550, and 16550A.  The "A" on the 16550A indicates that the
advanced options of the 16550 have been activated.
.TP
.BR baud_base " baud_base"
This option sets the base baud rate, which is the clock frequency divided
by 16.
.TP
.BR divisor " divisor"
This option sets the custom divison.  This divisor will be used then the
.B spd_cust
option is selected and the serial port is set to 38.4kb by the application.
.TP
.B sak
Set the break key at the Secure Attention Key
.TP
.B -sak
disable the Secure Attention Key
.TP
.B
spd_hi
Use 57.6kb when the application requests 38.4kb.
.TP
.B spd_vhi
Use 115kb when the application requests 38.4kb.
.TP
.B spd_cust
Use the custom divisor to set the speed when the application requests
38.4kb.  In this case, the baud rate is the
.B baud_base
divided by the
.BR divisor .
.TP
.B spd_normal
Use 38.4kb when the application requests 38.4kb.
.SH CAUTION
CAUTION: Using an invalid port can lock up your machine.
.SH BUGS
The "flags" information which
.B setserial
prints seems to be incorrect on some systems.
.SH FILES
.I /etc/rc.local
.SH "SEE ALSO"
.BR tty (4),
.BR ttys (4),
kernel/chr_drv/serial.c
.SH AUTHOR
Rick Sladkey (jrs@world.std.com)
.br
High speed modifications by: Michael K. Johnson (johnsonm@stolaf.edu)


usr/src/utile-1.1/setserial.c
/* setserial.c - get/set Linux serial port info - rick sladkey */
/* modified to do work again and added setting fast serial speeds,
   Michael K. Johnson, johnsonm@stolaf.edu */
/*
 * Very heavily modified --- almost rewritten from scratch --- to have
 * a more flexible command structure.  Now able to set any of the
 * serial-specific options using the TIOCSSERIAL ioctl().
 * 			Theodore Ts'o, tytso@mit.edu, 1/1/93
 */

#include <stdio.h>
#include <fcntl.h>
#include <termios.h>
#include <string.h>

#include <linux/fs.h>
#include <linux/tty.h>

char *progname;

struct serial_type_struct {
	int id;
	char *name;
} serial_type_tbl[] = {
	PORT_UNKNOWN,	"unknown",
	PORT_8250,	"8250",
	PORT_16450,	"16450",
	PORT_16550,	"16550",
	PORT_16550A,	"16550A",
	PORT_UNKNOWN,	"none",
	-1,		NULL
};

#define CMD_FLAG	1
#define CMD_PORT	2
#define CMD_IRQ		3
#define CMD_DIVISOR	4
#define CMD_TYPE	5
#define CMD_BASE	6

#define FLAG_CAN_INVERT	0x0001
#define FLAG_NEED_ARG	0x0002

struct flag_type_table {
	int	cmd;
	char	*name;
	int	bits;
	int	mask;
	int	flags;
} flag_type_tbl[] = {
	CMD_FLAG, 	"SAK", 		ASYNC_SAK, 	ASYNC_SAK, 	FLAG_CAN_INVERT,
	CMD_FLAG,	"FOURPORT",	ASYNC_FOURPORT, ASYNC_FOURPORT,	FLAG_CAN_INVERT,
	CMD_FLAG,	"SPD_NORMAL",	0,		ASYNC_SPD_MASK,		0,
	CMD_FLAG,	"SPD_HI",	ASYNC_SPD_HI, 	ASYNC_SPD_MASK, 0,
	CMD_FLAG,	"SPD_VHI",	ASYNC_SPD_VHI,	ASYNC_SPD_MASK,	0,
	CMD_FLAG,	"SPD_CUST",	ASYNC_SPD_CUST,	ASYNC_SPD_MASK,	0,
	CMD_PORT,	"port",		0,		0,		FLAG_NEED_ARG,
	CMD_IRQ,	"irq",		0,		0,		FLAG_NEED_ARG,
	CMD_DIVISOR,	"divisor",	0,		0,		FLAG_NEED_ARG,
	CMD_TYPE,	"uart",		0,		0,		FLAG_NEED_ARG,
	CMD_BASE,	"base",		0,		0,		FLAG_NEED_ARG,
	CMD_BASE,	"baud_base",	0,		0,		FLAG_NEED_ARG,
	0,		0,		0,		0,		0,
};
	
char *serial_type(int id)
{
	int i;

	for (i = 0; serial_type_tbl[i].id != -1; i++)
		if (id == serial_type_tbl[i].id)
			return serial_type_tbl[i].name;
	return "undefined";
}

int uart_type(char *name)
{
	int i;

	for (i = 0; serial_type_tbl[i].id != -1; i++)
		if (!strcasecmp(name, serial_type_tbl[i].name))
			return serial_type_tbl[i].id;
	return -1;
}


int atonum(char *s)
{
	int n;

	while (*s == ' ')
		s++;
	if (strncmp(s, "0x", 2) == 0 || strncmp(s, "0X", 2) == 0)
		sscanf(s + 2, "%x", &n);
	else if (s[0] == '0' && s[1])
		sscanf(s + 1, "%o", &n);
	else
		sscanf(s, "%d", &n);
	return n;
}

void print_flags(struct serial_struct *serinfo)
{
	struct	flag_type_table	*p;
	int	flags;

	flags = serinfo->flags;
	
	for (p = flag_type_tbl; p->name; p++) {
		if (p->cmd != CMD_FLAG)
			continue;
		if ((flags & p->mask) == p->bits)
			printf("%s ", p->name);
	}
	printf("\n");
}

void get_serial(char *device)
{
	struct serial_struct serinfo;
	int	fd;

	if ((fd = open(device, O_RDWR)) < 0) {
		perror(device);
		return;
	}
	if (ioctl(fd, TIOCGSERIAL, &serinfo) < 0) {
		perror("Cannot get serial info");
		close(fd);
		return;
	}
	if (serinfo.irq == 9)
		serinfo.irq = 2;	/* People understand 2 better than 9 */
	printf("%s, Line %d, UART: %s, Port: 0x%.4x, IRQ: %d, Baud_base: %d, Flags: ",
	       device, serinfo.line, serial_type(serinfo.type),
	       serinfo.port, serinfo.irq, serinfo.baud_base);
	print_flags(&serinfo);
	close(fd);
}

void set_serial(char *device, char ** arg)
{
	struct serial_struct old_serinfo, new_serinfo;
	struct	flag_type_table	*p;
	int	fd;
	int	do_invert = 0;
	char	*word;
	

	if ((fd = open(device, O_RDWR)) < 0) {
		perror(device);
		exit(1);
	}
	if (ioctl(fd, TIOCGSERIAL, &old_serinfo) < 0) {
		perror("Cannot get serial info");
		exit(1);
	}
	new_serinfo = old_serinfo;
	while (*arg) {
		do_invert = 0;
		word = *arg++;
		if (*word == '-') {
			do_invert++;
			word++;
		}
		for (p = flag_type_tbl; p->name; p++) {
			if (!strcasecmp(p->name, word))
				break;
		}
		if (!p->name) {
			fprintf(stderr, "Invalid flag: %s\n", word);
			exit(1);
		}
		if (do_invert && !(p->flags & FLAG_CAN_INVERT)) {
			fprintf(stderr, "This flag can not be inverted: %s\n", word);
			exit(1);
		}
		if ((p->flags & FLAG_NEED_ARG) && !*arg) {
			fprintf(stderr, "Missing argument for %s\n", word);
			exit(1);
		}
		switch (p->cmd) {
		case CMD_FLAG:
			new_serinfo.flags &= ~p->mask;
			if (!do_invert)
				new_serinfo.flags |= p->bits;
			break;
		case CMD_PORT:
			new_serinfo.port = atonum(*arg++);
			break;
		case CMD_IRQ:
			new_serinfo.irq = atonum(*arg++);
			break;
		case CMD_DIVISOR:
			new_serinfo.custom_divisor = atonum(*arg++);
			break;
		case CMD_TYPE:
			new_serinfo.type = uart_type(*arg++);
			if (new_serinfo.type < 0) {
				fprintf(stderr, "Illegal UART type: %s", *--arg);
				exit(1);
			}
			break;
		case CMD_BASE:
			new_serinfo.baud_base = atonum(*arg++);
			break;	
		default:
			fprintf(stderr, "Internal error: unhandled cmd #%d\n", p->cmd);
			exit(1);
		}
			
	}
	if (ioctl(fd, TIOCSSERIAL, &new_serinfo) < 0) {
		perror("Cannot set serial info");
		exit(1);
	}
}


void usage()
{
	fprintf(stderr,
		"usage: %s serial-device [opt1 [arg]] [opt2] ... \n\n", progname);
	fprintf(stderr, "Available options: (* = Takes an argument)\n");
	fprintf(stderr, "\t* port\t\tset the I/O port\n");
	fprintf(stderr, "\t* irq\t\tset the interrupt\n");	
	fprintf(stderr, "\t* uart\t\tset UART type (none, 8250, 16450, 16550, 16550A\n");
	fprintf(stderr, "\t* baud_base\tset base baud rate (CLOCK_FREQ / 16)\n");
	fprintf(stderr, "\t* divisor\tset the custom divisor (see below)\n");
	fprintf(stderr, "\t  sak\t\tset the break key as the Secure Attention Key\n");
	fprintf(stderr, "\t  -sak\t\tdisable the Secure Attention Key\n");
	fprintf(stderr, "\n");
	fprintf(stderr, "\t  spd_hi\tuse 56kb instead of 38.4kb\n");
	fprintf(stderr, "\t  spd_vhi\tuse 115kb instead of 38.4kb\n");
	fprintf(stderr, "\t  spd_cust\tuse the custom divisor to set the speed at 38.4kb\n");
	fprintf(stderr, "\t\t\t\t(baud rate = baud_base / custom_divisor)\n");
	fprintf(stderr, "\t  spd_normal\tuse 38.4kb when a buad rate of 38.4kb is selected\n");
	fprintf(stderr, "\n");
	fprintf(stderr, "Use a leading '0x' for hex numbers.\n");
	fprintf(stderr, "CAUTION: Using an invalid port can lock up your machine!\n");
	exit(1);
}

main(int argc, char **argv)
{
	progname = argv[0];
	if (argc == 1)
		usage();
	if (!strcmp(argv[1], "-g")) {
		argv += 2;
		while (*argv)
			get_serial(*argv++);
	} else if (argc == 2)
		get_serial(argv[1]);
	else
		set_serial(argv[1], argv+2);
	exit(0);
}

